File: algorithms/python/两两交换链表中的节点/两两交换链表中的节点.py
# Source : https://leetcode.cn/problems/swap-nodes-in-pairs/description/
# Author : Yanan Wang
# Date   : 2023-08-27

# 本题采用递归解法：
#    1 -> 2 -> 3 -> 4 -> 5 -> 6
#    对于以上1-6链表，我们可以先将3-6反转完毕，返回新的头节点 --- 1 -> 2 -> 4 -> 3 -> 6 -> 5 最后反转前两个结点，返回结果
#    对于以上3-6链表，我们可以先将4-6反转完毕，返回新的头节点 --- 3 -> 4 -> 6 -> 5 最后反转前两个结点，返回结果
#    那么重复子问题就是：给定n个结点的链表，先反转后面n-2个结点，最后反转前两个结点，返回结果
#    递归出口：当题目给的head为空时，返回空；当递归过程中发现当前节点的next为空时，返回当前节点。

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        newHead = head.next
        head.next = self.swapPairs(newHead.next)
        newHead.next = head
        return newHead

File: algorithms/cpp/两两交换链表中的节点/两两交换链表中的节点.cpp
// Source : https://leetcode.cn/problems/swap-nodes-in-pairs/description/
// Author : Yanan Wang
// Date   : 2023-10-02

/**
 * 做链表类型题目，不能吝啬创建变量
 * 针对本题，首先创建一个dummyHead，否则前面的结点和后面的结点的处理逻辑就不相同了
 *   ---        ---        ---        ---        ---
 *  | d |  ->  | 1 |  ->  | 2 |  ->  | 3 |  ->  | 4 |  
 *   ---        ---        ---        ---        ---
 *   prev       cur        next       nnext     
 *                          
 *                    ⬇    创建四个指针分别指向这四个结点，用这四个指针将前两个有效结点的关系调整完毕
 *
 *   ---        ---        ---        ---        ---
 *  | d |  ->  | 2 |  ->  | 1 |  ->  | 3 |  ->  | 4 |  
 *   ---        ---        ---        ---        ---
 *   prev       next       cur        nnext
 *
 *                    ⬇    调整完毕后，更新这四个指针，开始下一轮循环
 *
 *   ---        ---        ---        ---        ---
 *  | d |  ->  | 2 |  ->  | 1 |  ->  | 3 |  ->  | 4 |  
 *   ---        ---        ---        ---        ---
 *                         prev       cur        next        nnext
 */


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode *dummyHead = new ListNode();
        dummyHead->next = head;
        ListNode *prev = dummyHead;
        ListNode *cur = head;
        while (cur != nullptr && cur->next != nullptr) {
            ListNode *next = cur->next;
            ListNode *nnext = next->next;
            prev->next = next;
            next->next = cur;
            cur->next = nnext;
            prev = cur;
            cur = nnext;
        }
        return dummyHead->next;
    }
};

/**
 * 新增一种解法：递归法
*/
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        ListNode *node = swapPairs(head->next->next);
        ListNode *next = head->next;
        next->next = head;
        head->next = node;
        return next; 
    }
};